File: python/conifer_sdk/catalog.py
"""Reading the catalog: models, prices, and the cheapest model that declares
what a caller needs."""

from __future__ import annotations

from typing import Any, Mapping, Optional, Sequence

from .types import CatalogModel


def to_catalog_model(entry: Mapping[str, Any]) -> CatalogModel:
    """A catalog entry from ``GET /v1/models``. ``raw`` keeps every field."""
    return CatalogModel(
        id=str(entry.get("id", "")),
        endpoint_kind=entry.get("endpoint_kind"),
        display_name=entry.get("display_name"),
        provider=entry.get("provider"),
        context_window=entry.get("context_window"),
        max_output_tokens=entry.get("max_output_tokens"),
        max_tools=entry.get("max_tools"),
        caps=entry.get("caps"),
        embedding_dimensions=entry.get("embedding_dimensions"),
        pricing=entry.get("pricing"),
        fee_pct=entry.get("fee_pct"),
        unavailable=entry.get("unavailable"),
        raw=dict(entry),
    )


def price_of(model: CatalogModel) -> Optional[float]:
    """One comparable price per model: input plus three times output, in USD
    per million tokens. A ranking key, not a cost forecast. Catalog prices are
    decimal strings; an unrecognised pricing shape is unpriced, not free."""
    if not model.pricing:
        return None
    input_rate = _decimal(model.pricing.get("in_usd_per_mtok"))
    output_rate = _decimal(model.pricing.get("out_usd_per_mtok"))
    if input_rate is None and output_rate is None:
        return None
    return (input_rate or 0.0) + 3 * (output_rate or 0.0)


def _decimal(value: Any) -> Optional[float]:
    if isinstance(value, bool):
        return None
    if isinstance(value, (int, float)):
        return float(value)
    if not isinstance(value, str):
        return None
    try:
        return float(value)
    except ValueError:
        return None


def pick_cheapest(
    models: Sequence[CatalogModel],
    caps: Sequence[str] = (),
    min_context_window: Optional[int] = None,
) -> Optional[CatalogModel]:
    """The cheapest model that declares every capability in ``caps``. A model
    with no declared caps or no price is skipped, never assumed."""
    eligible = []
    for model in models:
        if model.unavailable:
            continue
        if min_context_window is not None:
            if model.context_window is None or model.context_window < min_context_window:
                continue
        if caps and (model.caps is None or not all(cap in model.caps for cap in caps)):
            continue
        if price_of(model) is None:
            continue
        eligible.append(model)
    if not eligible:
        return None
    return min(eligible, key=lambda m: price_of(m) or 0)

File: src/catalog.ts
// Reading the catalog: models, prices, and the cheapest model that declares
// what a caller needs.

import type { CatalogModel } from "./types.ts";

/** A catalog entry from `GET /v1/models`. `raw` keeps every field. */
export function toCatalogModel(entry: Record<string, unknown>): CatalogModel {
  return {
    id: String(entry.id ?? ""),
    endpointKind: entry.endpoint_kind as string | undefined,
    displayName: entry.display_name as string | undefined,
    provider: entry.provider as string | undefined,
    contextWindow: entry.context_window as number | undefined,
    maxOutputTokens: entry.max_output_tokens as number | undefined,
    minOutputTokens: entry.min_output_tokens as number | undefined,
    outputTokenLimitSupported: entry.output_token_limit_supported as boolean | undefined,
    maxTools: entry.max_tools as number | undefined,
    caps: entry.caps as string[] | undefined,
    embeddingDimensions: entry.embedding_dimensions as number | undefined,
    pricing: entry.pricing as CatalogModel["pricing"],
    feePct: entry.fee_pct as number | undefined,
    unavailable: entry.unavailable as boolean | undefined,
    raw: entry,
  };
}

/**
 * The cheapest model that declares every capability in `caps`. A model with
 * no declared caps, no price, or explicit inability to honor output limits is skipped.
 */
export function pickCheapest(
  models: CatalogModel[],
  caps: string[],
  options: { minContextWindow?: number } = {},
): CatalogModel | undefined {
  const eligible = models.filter((model) => {
    if (model.unavailable === true) return false;
    if (model.outputTokenLimitSupported === false) return false;
    if (options.minContextWindow !== undefined) {
      if (model.contextWindow === undefined || model.contextWindow < options.minContextWindow) return false;
    }
    if (caps.length === 0) return true;
    if (model.caps === undefined) return false;
    return caps.every((cap) => model.caps?.includes(cap));
  });
  const priced = eligible.filter((model) => priceOf(model) !== undefined);
  priced.sort((a, b) => (priceOf(a) as number) - (priceOf(b) as number));
  return priced[0];
}

/**
 * One comparable price per model: input plus three times output, in USD per
 * million tokens. A ranking key, not a cost forecast. Catalog prices are
 * decimal strings; an unrecognised pricing shape is unpriced, not free.
 */
export function priceOf(model: CatalogModel): number | undefined {
  const pricing = model.pricing;
  if (pricing === undefined) return undefined;
  const input = decimal(pricing.in_usd_per_mtok);
  const output = decimal(pricing.out_usd_per_mtok);
  if (input === undefined && output === undefined) return undefined;
  return (input ?? 0) + 3 * (output ?? 0);
}

function decimal(value: unknown): number | undefined {
  if (typeof value === "number" && Number.isFinite(value)) return value;
  if (typeof value !== "string") return undefined;
  const parsed = Number.parseFloat(value);
  return Number.isFinite(parsed) ? parsed : undefined;
}
